src/worker.py

```
@@ -1697,6 +1697,301 @@ def _persist_attachment(self, session_id, info):
+        user_id = info["user_id"]
+        self.sessions.pop(sid, None)
+
+        still_connected = any(s["user_id"] == user_id for s in self.sessions.values())
+        if not still_connected:
+            self.presence.pop(user_id, None)
+            self._broadcast(json.dumps({"type": "leave", "user_id": user_id}))
+
+    async def on_webSocketError(self, ws, error):
+        print(f"[PresenceDO.on_webSocketError] error={error!r}")
+
+    def _send_welcome(self, ws, session_id, user_id):
```

Ruff (0.15.10)

[warning] Missing return type annotation for private function `_send_welcome`

Add return type annotation: `None`

(ANN202)